### PR TITLE
Initialize repo and add simple memory tests

### PR DESCRIPTION
**Pull request by d33bs**
This PR adds content to help initialize the repo along with simple examples of testing memory allocation reference counts (where nothing unusual is expected). It builds a scaffold where additional tests and further specification could be provided. As mentioned in the `README.md`, the work is largely inspired by and may help inform progress for cytomining/CytoTable#75 (in addition to other areas). As a background, I opted for containerized testing of these circumstances to help isolate what I noticed as differences between MacOS Apple Silicon and other OS's when it came to memory management in Python.

Wondering aloud: perhaps the word `test` is too overloaded and moving to `experiment` for memory testing would be better?

Thank you in advance for any thoughts and feedback you may have! Please don't hesitate to let me know if you have any questions.

CC @chrisroederucdenver

### CODE REVIEW

**Comment by chrisroederucdenver** [comment]
I'm not seeing much in the README as mentioned in the e-mail. 
(Right branch, wrong repo. I was seeing the one liner.  FIixed that.  I see the "#Development" paragraph. It could use tweaks for the commands to install as I list below, YMMV.

Browsing the code, I could easily enough see the pipeline and Docker etc and distinguish them from the tests. Maybe you missed a commit.
Perhaps obviously, a few lines describing what you're testing, the different gc modules and the different ways to call them would be really useful. 

Being much newer to Python, some basic "how to get started" steps would help.

tl;dr You need Docker running and some stuff installed after python3: (brew) python, (brew) pipx, (pipx) poetry and (pipx) poethepoet. Load the depdendencis with "poetry install", the run by doing "poe run_all_tests". 

Here's the travelogue I promised: 
- brew install pipx
- pipx ensurepath
  - d'oh brew failed b/c no command line tools
- pipx install poetry
- poetry run src/pymaccounter/runner.py 
  - permisison denied, no she-bang?
  - /Users/roederc/.local/bin/poetry run python3 src/pymaccounter/runner.py
  - no module named 'anyio'
  - hmmm, view pyproject.toml. What's this poet-the-poet thing in the pyproject.toml?? .
    -  I know that picture, oh, Poe the Poet
  - pipx install poethepoet
- poe run_all_tests
  - No module named 'anyio', oops
- poetry install
- poe run_all_tests
- Cannot connect to the Docker daemon
  - click the icon. Oh look at those containers from a year ago...
-  poe run_all_tests

Tah-Dah: 
0: [0.40s] gc: collectable <tuple 0xffff9f8f7ee0>
30: exec poetry run python src/pymaccounter/tests/test_multiply_gc.debug_leak.py DONE

20: exec poetry run python src/pymaccounter/tests/test_baseline.py DONE
20: exec poetry run python src/pymaccounter/tests/test_baseline.py DONE

33: exec poetry run python src/pymaccounter/tests/test_multiply_gc.collect.py DONE

File: src/pymaccounter/tests/test_multiply_gc.callbacks.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# collect garbage
+# see here for more info:
+# https://docs.python.org/3/library/gc.html#gc.collect
+collected = gc.collect()
+
+# print the unreachable objects
+print(collected)
+
+# print the len of gc garbage to reference unreachable objects
+print(len(gc.garbage))
```
Comment (chrisroederucdenver) [inline src/pymaccounter/tests/test_multiply_gc.callbacks.py]: As a test, I presume you could check this against zero as you expect?

File: src/pymaccounter/tests/test_multiply_gc.collect.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# collect garbage
+# see here for more info:
+# https://docs.python.org/3/library/gc.html#gc.collect
+collected = gc.collect()
+
+# print the unreachable objects
+print(collected)
+
+# print the len of gc garbage to reference unreachable objects
+print(len(gc.garbage))
```
Comment (chrisroederucdenver) [inline src/pymaccounter/tests/test_multiply_gc.collect.py]: As above,  I presume you could check this against zero as you expect?

File: src/pymaccounter/tests/test_multiply_gc.debug_leak.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# collect garbage
+# see here for more info:
+# https://docs.python.org/3/library/gc.html#gc.collect
+collected = gc.collect()
+
+# print the unreachable objects
+print(collected)
+
+# print the len of gc garbage to reference unreachable objects
+print(len(gc.garbage))
```
Comment (chrisroederucdenver) [inline src/pymaccounter/tests/test_multiply_gc.debug_leak.py]: As above,  I presume you could check this against zero as you expect?

File: src/pymaccounter/tests/test_multiply_gc.callbacks.py (hunk 1 above)
Comment (d33bs) [reply src/pymaccounter/tests/test_multiply_gc.callbacks.py]: Thanks Chris! This is a fantastic idea, and I've moved to use asserts in later commits.

File: src/pymaccounter/tests/test_multiply_gc.collect.py (hunk 2 above)
Comment (d33bs) [reply src/pymaccounter/tests/test_multiply_gc.collect.py]: Thanks again, moved to assertions / a test-driven approach in later commits.

File: src/pymaccounter/tests/test_multiply_gc.debug_leak.py (hunk 3 above)
Comment (d33bs) [reply src/pymaccounter/tests/test_multiply_gc.debug_leak.py]: Thanks again, moved to assertions / a test-driven approach in later commits.

**Comment by d33bs** [comment]
Thank you so much @chrisroederucdenver for the review and thoughts! I've added a number of changes to this work based on your comments here and our various discussions. I've moved this to use a test-driven approach within the repo as you suggested (thanks for this, fantastic idea!).

I plan to expand on the existing tests with additional future PR's focused on specific aspects (mostly focusing on building a scaffold / proof-of-concept for now). Mostly this is to limit the scope of additions given the number of possible variables at play (Python version, interpreter version, OS, memory observation tools used, etc).

Your comments also made me feel that further documentation was warranted surrounding memory and how Python manages memory. I've moved this content to CU-DBMI/set-website#28 which will result in a blog post surrounding the topic.

More specific comments:

- Getting started: I added greater documentation surrounding how to get started with concerns to this repo's use of Python.
- More docs on tests: I added some documentation covering this and outlining how test documentation could function for this repo.
- Missing hashbangs: generally these are considered optional for Python, esp. with concerns to packages (they're mostly used in relationship to scripts instead of modules/packages). My own style tends to avoid the use of these as I seek to avoid maintaining an additional line of code per file which may void the practice of DRY.

**Comment by d33bs** [comment]
Thanks again @chrisroederucdenver, appreciate your thoughts on this throughout! Merging this for now with an eye towards further improvements in the future.
